File: analyzers/tests/SonarAnalyzer.Test/TestCases/ThrowReservedExceptions.vb
Public Class ThrowReservedExceptions
    Public Sub Method1()
        Throw New Exception()                   ' Noncompliant {{'System.Exception' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^
        Throw New ApplicationException()        ' Noncompliant {{'System.ApplicationException' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^^^^^^^^^^^^
        Throw New SystemException()             ' Noncompliant {{'System.SystemException' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^^^^^^^
        Throw New ExecutionEngineException()    ' Noncompliant {{'System.ExecutionEngineException' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        Throw New IndexOutOfRangeException()    ' Noncompliant {{'System.IndexOutOfRangeException' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        Throw New NullReferenceException()      ' Noncompliant {{'System.NullReferenceException' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        Throw New OutOfMemoryException()        ' Noncompliant {{'System.OutOfMemoryException' should not be thrown by user code.}}
        '     ^^^^^^^^^^^^^^^^^^^^^^^^^^
        Dim e = New OutOfMemoryException()  ' Compliant
        Throw New ArgumentNullException()   ' Compliant

        Try
            Dim a = New Integer(-1) {}
            ' Throw exception
            Console.WriteLine(a(1))
        Catch generatedExceptionName As IndexOutOfRangeException ' Compliant
            Throw
        End Try
    End Sub
End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ThrowReservedExceptions.cs
using System;

public class ThrowReservedExceptions
{
    public void Method1()
    {
        throw new Exception();                  // Noncompliant {{'System.Exception' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^

        throw new ApplicationException();       // Noncompliant {{'System.ApplicationException' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^^^^^^^^^^^^

        throw new SystemException();            // Noncompliant {{'System.SystemException' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^^^^^^^

        throw new ExecutionEngineException();   // Noncompliant {{'System.ExecutionEngineException' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

        throw new IndexOutOfRangeException();   // Noncompliant {{'System.IndexOutOfRangeException' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

        throw new NullReferenceException();     // Noncompliant {{'System.NullReferenceException' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

        throw new OutOfMemoryException();       // Noncompliant {{'System.OutOfMemoryException' should not be thrown by user code.}}
//            ^^^^^^^^^^^^^^^^^^^^^^^^^^

        var e = new OutOfMemoryException(); // Compliant
        throw new ArgumentNullException();  // Compliant

        OutOfMemoryException e1 = (OutOfMemoryException)new ArgumentException(); // Error [CS0030] - cannot cast

        try
        {
            var a = new int[0];
            Console.WriteLine(a[1]); // Throw exception
        }
        catch (IndexOutOfRangeException)
        {
            throw; // Compliant
        }
    }

    public void Arrow() =>
        throw new Exception();          // Noncompliant

    public Action Lambda() =>
        () => throw new Exception();    // Noncompliant
}
